palace/linalg/rap.cpp

```diff
@@ -0,0 +1,809 @@
+  if (auto *bfA = dynamic_cast<const mfem::BilinearForm *>(A))
+  {
+    if (bfA->HasSpMat())
+    {
+      bfA->SpMat().GetDiag(ly);
+    }
+    else if (bfA->HasExt())
+    {
+      bfA->Ext().AssembleDiagonal(ly);
+    }
+    else
+    {
+      MFEM_ABORT("Unable to assemble the local operator diagonal of BilinearForm!");
+    }
+  }
+  else if (auto *sA = dynamic_cast<const mfem::SparseMatrix *>(A))
+  {
+    sA->GetDiag(ly);
+  }
+  else
+  {
+    MFEM_ABORT("ParOperator::AssembleDiagonal requires A as a BilinearForm or "
+               "SparseMatrix!");
+  }
+
+  // Parallel assemble and eliminate essential true dofs.
+  const Operator *P = test_fespace.GetProlongationMatrix();
+  if (const auto *hP = dynamic_cast<const mfem::HypreParMatrix *>(P))
+  {
+    hP->AbsMultTranspose(1.0, ly, 0.0, diag);
+  }
+  else
+  {
+    P->MultTranspose(ly, diag);
+  }
+  if (dbc_tdof_list)
+  {
+    if (diag_policy == DiagonalPolicy::DIAG_ONE)
+    {
+      linalg::SetSubVector(diag, *dbc_tdof_list, 1.0);
+    }
+    else if (diag_policy == DiagonalPolicy::DIAG_ZERO)
+    {
+      linalg::SetSubVector(diag, *dbc_tdof_list, 0.0);
+    }
+  }
+}
+
+mfem::HypreParMatrix &ParOperator::ParallelAssemble()
+{
+  if (RAP)
+  {
+    return *RAP;
+  }
+
+  // XX TODO: For mfem::AssemblyLevel::PARTIAL, we cannot use CeedOperatorFullAssemble for
+  //          a ND space with p > 1. We should throw an error here that the user needs to
+  //          use AssemblyLevel::LEGACY in this case.
+
+  // Build the square or rectangular RAP HypreParMatrix.
+  if (&trial_fespace == &test_fespace)
+  {
+    mfem::SparseMatrix *lA;
+    bool own_lA = false;
+    if (auto *bfA = dynamic_cast<mfem::BilinearForm *>(A))
+    {
+#ifdef MFEM_USE_CEED
+      if (bfA->HasSpMat())
+      {
+        lA = &bfA->SpMat();
+      }
+      else if (bfA->HasExt())
+      {
+        lA = mfem::ceed::CeedOperatorFullAssemble(*bfA);
+        own_lA = true;
+      }
+      else
+      {
+        MFEM_ABORT("Unable to assemble the local operator for parallel assembly of "
+                   "BilinearForm!");
+        lA = nullptr;
+      }
+#else
+      MFEM_VERIFY(bfA->HasSpMat(),
+                  "Missing assembled SparseMatrix for parallel assembly of BilinearForm!");
+      lA = &bfA->SpMat();
+#endif
+    }
+    else if (auto *sA = dynamic_cast<mfem::SparseMatrix *>(A))
+    {
+      lA = sA;
+    }
+    else
+    {
+      MFEM_ABORT("ParOperator::ParallelAssemble requires A as a BilinearForm or "
+                 "SparseMatrix!");
+      lA = nullptr;
+    }
+    mfem::HypreParMatrix *hA =
+        new mfem::HypreParMatrix(trial_fespace.GetComm(), trial_fespace.GlobalVSize(),
+                                 trial_fespace.GetDofOffsets(), lA);
+    const mfem::HypreParMatrix *P = trial_fespace.Dof_TrueDof_Matrix();
+    RAP = std::make_unique<mfem::HypreParMatrix>(hypre_ParCSRMatrixRAP(*P, *hA, *P), true);
+    delete hA;
```

Comment:
This works because of the way the cast is defined in the MFEM wrapper:

```
/// Typecasting to hypre's hypre_ParCSRMatrix*
operator hypre_ParCSRMatrix*() const { return A; }
```